doc/source/installation.rst

```diff
@@ -42,3 +42,184 @@ master branch). To install these wheels, run the following command:
 .. _`MacOS Python 3.6`: https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev3-cp36-cp36m-macosx_10_6_intel.whl
 .. _`MacOS Python 3.5`: https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev3-cp35-cp35m-macosx_10_6_intel.whl
 .. _`MacOS Python 2.7`: https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev3-cp27-cp27m-macosx_10_6_intel.whl
+
+Building Ray from Source
+------------------------
+
+Installing from ``pip`` should be sufficient for most Ray users.
+
+However, should you need to build from source, follow instructions below for both Linux and MacOS.
+
+Dependencies
+~~~~~~~~~~~~
+
+To build Ray, first install the following dependencies. We recommend using
+`Anaconda`_.
+
+.. _`Anaconda`: https://www.continuum.io/downloads
+
+For Ubuntu, run the following commands:
+
+.. code-block:: bash
+
+  sudo apt-get update
+  sudo apt-get install -y build-essential curl unzip psmisc
+
+  # If you are not using Anaconda, you need the following.
+  sudo apt-get install python-dev  # For Python 2.
+  sudo apt-get install python3-dev  # For Python 3.
+
+  pip install cython==0.29.0
+
+For MacOS, run the following commands:
+
+.. code-block:: bash
+
+  brew update
+  brew install wget
+
+  pip install cython==0.29.0
+
+
+If you are using Anaconda, you may also need to run the following.
+
+.. code-block:: bash
+
+  conda install libgcc
+
+
+Install Ray
+~~~~~~~~~~~
+
+Ray can be built from the repository as follows.
+
+.. code-block:: bash
+
+  git clone https://github.com/ray-project/ray.git
+
+  # Install Bazel.
+  ray/ci/travis/install-bazel.sh
+
+  cd ray/python
+  pip install -e . --verbose  # Add --user if you see a permission denied error.
+
+Alternatively, Ray can be built from the repository without cloning using pip.
+
+.. code-block:: bash
+
+    pip install git+https://github.com/ray-project/ray.git#subdirectory=python
+
+Cleaning the source tree
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The source tree can be cleaned by running
+
+.. code-block:: bash
+
+  git clean -f -f -x -d
+
+in the ``ray/`` directory. Warning: this command will delete all untracked files
+and directories and will reset the repository to its checked out state.
+For a shallower working directory cleanup, you may want to try:
+
+.. code-block:: bash
+
+  rm -rf ./build
+
+under ``ray/``. Incremental builds should work as follows:
+
+.. code-block:: bash
+
+  pushd ./build && make && popd
+
+under ``ray/``.
+
+Docker Source Images
```

Comment:
Not sure if this section belongs under "installation", though I don't feel strongly.